File: binary-trees/binary-tree-right-side-view/binary-tree-right-side-view.go
package binarytreerightsideview

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func rightSideView(root *TreeNode) []int {
	list := []int{}
	traverse(root, 0, &list)
	return list
}

func traverse(root *TreeNode, level int, list *[]int) {
	if root == nil {
		return
	}

	if len(*list) == level {
		*list = append(*list, root.Val)
	}

	traverse(root.Right, level+1, list)
	traverse(root.Left, level+1, list)
}

File: binary-trees/binary-tree-right-side-view/binary-tree-right-side-view.js
function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}

/**
 * @param {TreeNode} root
 * @return {number[]}
 */
function rightSideView(root) {
  const list = [];
  traverse(root, 0, list);
  return list;
}

function traverse(root, level, list) {
  if (root === null) {
    return;
  }

  if (list.length === level) {
    list.push(root.val);
  }

  traverse(root.right, level + 1, list);
  traverse(root.left, level + 1, list);
}
